File: ios/Packages/GemstoneServices/Sources/Stores/NotificationStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemNotificationStore
import struct Gemstone.InAppNotification
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneNotificationStore: GemNotificationStore, @unchecked Sendable {
    private let store: InAppNotificationStore

    public init(store: InAppNotificationStore) {
        self.store = store
    }

    public func saveNotifications(notifications: [Gemstone.InAppNotification]) async throws {
        try store.addNotifications(notifications.map { $0.toPrimitives() })
    }

    public func hasUnreadNotifications(walletId: String) async throws -> Bool {
        try store.hasUnreadNotifications(walletId: WalletId.from(id: walletId))
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/NotificationStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.InAppNotificationsDao
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.ext.toPrimitives
import uniffi.gemstone.GemNotificationStore

class GemstoneNotificationStore(private val notificationsDao: InAppNotificationsDao) : GemNotificationStore {

    override suspend fun saveNotifications(notifications: List<uniffi.gemstone.InAppNotification>) = notificationsDao.put(notifications.map { it.toPrimitives().toRecord() })

    override suspend fun markNotificationsRead(walletId: String, createdBefore: Long) = notificationsDao.markRead(walletId, createdBefore, System.currentTimeMillis())

    override suspend fun hasUnreadNotifications(walletId: String): Boolean = notificationsDao.hasUnread(walletId)
}
